sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py

```diff
@@ -1000,25 +1000,151 @@ def make_pack_name(names):
         *[
             core.CombineFn.from_runner_api(combine_payload.combine_fn, context)  # type: ignore[arg-type]
             for combine_payload in combine_payloads
-        ]).to_runner_api(context)  # type: ignore[arg-type]
+        ])  # type: ignore[arg-type]
     pack_transform = beam_runner_api_pb2.PTransform(
-        unique_name=pack_transform_name + '/Pack',
+        unique_name=pack_transform_name + '/CombinePerKey',
         spec=beam_runner_api_pb2.FunctionSpec(
             urn=common_urns.composites.COMBINE_PER_KEY.urn,
             payload=beam_runner_api_pb2.CombinePayload(
-                combine_fn=pack_combine_fn,
+                combine_fn=pack_combine_fn.to_runner_api(context),
                 accumulator_coder_id=tuple_accumulator_coder_id).
             SerializeToString()),
         inputs={'in': input_pcoll_id},
         # 'None' single output key follows convention for CombinePerKey.
         outputs={'None': pack_pcoll_id},
         environment_id=fused_stage.environment)
     pack_stage = Stage(
-        pack_stage_name + '/Pack', [pack_transform],
+        pack_stage_name + '/CombinePerKey', [pack_transform],
         downstream_side_inputs=fused_stage.downstream_side_inputs,
         must_follow=fused_stage.must_follow,
         parent=fused_stage.parent,
         environment=fused_stage.environment)
+
+    # Traverse the subtransform structure.
+    original_group_by_key_transforms = []
+    original_combine_grouped_values_transforms = []
+    original_combine_values_par_do_transforms = []
+    for transform in transforms:
```

Comment:
This whole thing looks pretty brittle. Can we not remove sub-transforms from the CombinePerKey transform in the first place (before generating the proto) so that we do not manually have to modify the proto here ?

Reply:
We would still have to insert the multiplexing DoFn as a follow-on. 

I do agree this function is getting a bit unwieldy though. 

Could we perhaps factor this out into a function `create_combine_per_key_transform(combine_fn, input_pcoll, output_pcoll, template)` that also creates the internal structure?